terraform/context.go

```diff
@@ -892,7 +906,15 @@ func (c *Context) walk(
 	return walker, realErr
 }
 
-func (c *Context) watchStop(walker *ContextGraphWalker, stopCh, doneCh <-chan struct{}) {
+func (c *Context) watchStop(walker *ContextGraphWalker, doneCh <-chan struct{}) {
+	// Get the stop channel. runContext might be nil only during tests.
+	// If this is called during a proper run operation, this will never
+	// be nil.
+	var stopCh <-chan struct{}
+	if ctx := c.runContext; ctx != nil {
```

Comment:
yes, but `watchStop` doesn't take a lock, and can race with any method writing to the context fields. Just verified this shows up readily with the race detector.

This should fix it by removing the read of that field from watchStop entirely:

```
@@ -811,7 +811,9 @@ func (c *Context) walk(

        // Watch for a stop so we can call the provider Stop() API.
        doneCh := make(chan struct{})
-       go c.watchStop(walker, doneCh)
+       stopCh := c.runContext.Done()
+
+       go c.watchStop(walker, stopCh, doneCh)

        // Walk the real graph, this will block until it completes
        realErr := graph.Walk(walker)
@@ -906,12 +908,7 @@ func (c *Context) walk(
        return walker, realErr
 }

-func (c *Context) watchStop(walker *ContextGraphWalker, doneCh <-chan struct{}) {
-       // Get the stop channel. runContext will never be nil since this should
-       // only be called within the context of an operation started with
-       // acquireRun
-       stopCh := c.runContext.Done()
-
+func (c *Context) watchStop(walker *ContextGraphWalker, stopCh, doneCh <-chan struct{}) {
        // Wait for a stop or completion
        select {
        case <-stopCh:
```